### PR TITLE
Improve performance of batch request worker

### PR DESCRIPTION
**Pull request by hugecannon**
**Fixes** # (*issue*)

I'm not sure exactly which time complexity the original code falls under (e.g. O(n^2) ), but the problem only becomes apparent with larger batch sizes. Batches larger than ~ 400 events were causing severe performance issues.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

[K6](https://k6.io/) results locally with batch size of 400

```
  before 

  scenarios: (100.00%) 1 scenario, 200 max VUs, 1m30s max duration (incl. graceful stop):
           * constant_request_rate: 2.83 iterations/s for 1m0s (maxVUs: 100-200, gracefulStop: 30s)


    iteration_duration.........: avg=616.1ms  min=575.44ms med=607.48ms max=907.34ms p(90)=643.99ms p(95)=663.54ms


  after


  scenarios: (100.00%) 1 scenario, 200 max VUs, 1m30s max duration (incl. graceful stop):
           * constant_request_rate: 2.83 iterations/s for 1m0s (maxVUs: 100-200, gracefulStop: 30s)


    iteration_duration.........: avg=77.19ms  min=43.47ms  med=78.15ms  max=130.14ms p(90)=95.7ms   p(95)=100.47ms
```

Batch sizes much larger than that cause request build up and eventual failed requests.

My K6 test script:

```
import http from 'k6/http';


let file400 = open('./400'); // request body produced by using  `go run genload.go -n 400`

export let options = {
  scenarios: {
    constant_request_rate: {
      executor: 'constant-arrival-rate',
      rate: 170,
      timeUnit: '1m',
      duration: '5m',
      preAllocatedVUs: 25, // how large the initial pool of VUs would be
      maxVUs: 200, // if the preAllocatedVUs are not enough, we can initialize more
    },
  },
};

export default function () {

  var url = 'http://MY_WRITE_KEY:@localhost:8080/v1/batch';

  var payload = file400;

  var params = {
    headers: {
      'Content-Type': 'application/json',
      'Accept-Encoding': 'gzip',
    },
  };

  http.post(url, payload, params);
}
```

### CODE REVIEW

**Comment by hugecannon** [comment]
This PR will need some help from a testing point of view
I couldn't find any tests covering this exact scenario and I'm not too familiar with the testing framework.

I did a manual check on the output of the function and it seemed to be the same given my limited inputs.

**Review by chandumlg — APPROVED** [review]

File: gateway/gateway.go [hunk 1]
```diff
@@ -383,14 +385,20 @@ func (gateway *HandleT) userWebRequestWorkerProcess(userWebRequestWorker *userWe
...
-					body, _ = sjson.SetBytes(body, fmt.Sprintf(`batch.%v.messageId`, index), uuid.NewV4().String())
+
+				toSet := vjson.Value().(map[string]interface{})
+				toSet["rudderId"] = rudderId
+				if messageId := strings.TrimSpace(vjson.Get("messageId").String()); messageId == "" {
+					toSet["messageId"] = uuid.NewV4().String()
 				}
+				out = append(out, toSet)
+
 				index++
```
Comment (sivashanmukh) [inline gateway/gateway.go:396]: Is this still needed?

Comment (hugecannon) [reply gateway/gateway.go:396]: Good point

**Review by sivashanmukh — APPROVED** [review]
LGTM

**Comment by chandumlg** [comment]
Thank you very much for the contribution.

**Pull request merged**